{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO\n",
    "* Make `add_rna_seq_data` and `create_or_update` more efficient\n",
    "* Add arabidopsis orthologues to DB\n",
    "  * Get from accession numbers (NIH and )\n",
    "* Connect front end to db for RNA-seq\n",
    "  * search by:\n",
    "    * At gene\n",
    "    * [X] Xe gene\n",
    "    * [X] GO terms\n",
    "  * Select experiment\n",
    "* add species.id to experiments model so that i can filter by exp for specific species\n",
    "\n",
    "\n",
    "* Maybe clear the textbox on expression_page when user changes something in sidebar and add text in box physically rather than placeholder\n",
    "* display genes for different GO terms in different panels?\n",
    "  \n",
    "## Done\n",
    "* Show which queried Xe genes arent in DB\n",
    "* plot expression values for Xerophyta genes\n",
    "  * log or normalised\n",
    "* Query expression data by experiment in DB (currently searches all of expression table)\n",
    "* Fix button stopping showing raw data\n",
    "* add DEGs table and DEG data\n",
    "  * filter expression data by DEGs\n",
    "* Fix up `gene_query_page.py`\n",
    "  * move the functions to the database\n",
    "  \n",
    "## Questions\n",
    "- How should i handle the clusters because they aren't consistent?\n",
    "- I need the humilis fasta file from nicci\n",
    "- Cant add schlecteri seedlings RNA seek data because idk what half the columns are\n",
    "- One of mamosas files (Xs) has gene counts all of 0\n",
    "- what are mamosas files\n",
    "- Michaels adult leaf tissues dont have any info in the columns\n",
    "- Does it make sense that X humilis has 99000 genes and other two have 28000\n",
    "- The At mapping has a lot of unplaced scaffold         unknown protein  \n",
    "  - any idea what these are? should they be treated to no blast hit (assumed yes)?\n",
    "- X elegans has 27920 genes in annotaition file but 28572 genes in FASTA, why?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01-01-2025 Add RNAseq data to DB\n",
    "Need to add the X. elegans RNA-seq data to the db. \n",
    "\n",
    "Steps:\n",
    "1. Add tables to `models.py`  and add info to DB schema with Alembic  \n",
    "2. Format data into long data format  \n",
    "\n",
    "### 1. Add table\n",
    "**Set up alembic** \n",
    "\n",
    "`alembic init alembic`  \n",
    "... chatGPT for other steps  \n",
    "\n",
    "**Create a migration script**   \n",
    "\n",
    "`alembic revision --autogenerate -m \"Your migration description here\"`  \n",
    "fill in script\n",
    "`alembic upgrade head`  \n",
    "\n",
    "\n",
    "Added two tables: `Gene_expressions` and `Experiments`\n",
    "\n",
    "### 2. Format the data\n",
    "\n",
    "Use the `data_tidier.py` script and the below python cells\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import data_tidier as dt\n",
    "import pandas as pd\n",
    "\n",
    "# Load the RNA expression file\n",
    "file_path = \"all_data/Michael_RNAseq/Xs seedlings/Xs_seedlings_DESeq2 normalised counts table.csv\"\n",
    "df = pd.read_csv(file_path)\n",
    "\n",
    "# Transform the data to long formathead \n",
    "long_df = dt.transform_to_long(df)\n",
    "\n",
    "long_df.to_csv(f\"{file_path}_tidy\", index=False)\n",
    "print(long_df.head())\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = dt.add_log2(long_df) # add log2 transformation\n",
    "df = dt.format_time_points(df) # format timepoint\n",
    "df.rename(columns={\"expression\": \"normalised_expression\"}, inplace=True)\n",
    "print(df.head())\n",
    "\n",
    "df.to_csv(\"all_data/Michael_RNAseq/Xe_seedlings (updated)/Xe_seedlings_DESeq2_normalised_counts_table_tidy_for_db.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Add to database\n",
    "Added X. elegans seedling time course data.  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import db_manager as dbm\n",
    "\n",
    "\n",
    "species_id = \"X. elegans\"\n",
    "experiment_name = \"xe_seedlings_time_course\"\n",
    "rna_seq_data = pd.read_csv(\"all_data/Michael_RNAseq/Xe_seedlings (updated)/Xe_seedlings_DESeq2_normalised_counts_table_tidy_for_db.csv\")\n",
    "\n",
    "dbm.add_experiment(experiment_name, \"time course of X. elegans seedlings\")\n",
    "dbm.add_rna_seq_data(rna_seq_data, species_id, experiment_name )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 02.01.2025 Connect frontend to DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import db_manager as dbm\n",
    "import db\n",
    "\n",
    "species_id = \"X. elegans\"\n",
    "experiment_name = \"xe_seedlings_time_course\"\n",
    "database = db.DB()\n",
    "\n",
    "database.link_experiment_to_species(experiment_name, species_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5-01-2025 Add DEG table\n",
    "\n",
    "- added DEG tables and data for X elegans seedlings.\n",
    "- can filter by DEGs now\n",
    "- Can also query expression data based on GO terms\n",
    "\n",
    "#### Testing DEG implementation\n",
    "\"12\",\"Xele.ptg000001l.15\",\"ReT48\",\"Up-regulated\",\"None\",\"None\"\n",
    "\"46\",\"Xele.ptg000001l.50\",\"None\",\"None\",\"DeT09\",\"Down-regulated\"\n",
    "\"137\",\"Xele.ptg000001l.159\",\"ReT04\",\"Up-regulated\",\"DeT12\",\"Down-regulated\"\n",
    "\n",
    "Xele.ptg000001l.15\n",
    "Xele.ptg000001l.50\n",
    "Xele.ptg000001l.159\n",
    "\n",
    "nice example: nuclear ubiquitin ligase complex\n",
    "  - 12 genes\n",
    "  - 6 degs\n",
    "  - 3 up \n",
    "  - 3 down"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import db \n",
    "\n",
    "database = db.DB()\n",
    "\n",
    "gene = \"Xsch.v2.MJHO01000001.1.1119\"\n",
    "genes =database.get_genes_by_go_term_or_description([\"jasmonic acid and ethylene-dependent systemic resistance\"], \"X. elegans\")\n",
    "gene_names = [gene.gene_name for gene in genes]\n",
    "\n",
    "print(len(gene_names))\n",
    "\n",
    "data = database.get_gene_expression_data(gene_names, \"xe_seedlings_time_course\")\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import models\n",
    "from sqlalchemy import func\n",
    "\n",
    "def get_go_terms_with_fewest_genes(threshold=0):\n",
    "    \"\"\"\n",
    "    Query GO terms linked to the fewest genes in ascending order of gene count.\n",
    "\n",
    "    Parameters:\n",
    "        session: SQLAlchemy session object.\n",
    "\n",
    "    Returns:\n",
    "        List of tuples containing GO term ID, GO term name, and gene count.\n",
    "    \"\"\"\n",
    "    database = db.DB()\n",
    "    session = database.session\n",
    "    query = (\n",
    "        session.query(\n",
    "            models.GO.go_id,          # Select the GO ID\n",
    "            models.GO.go_name,        # Select the GO Name\n",
    "            func.count(models.Gene.id).label(\"gene_count\")  # Count distinct genes\n",
    "        )\n",
    "        .join(models.annotations_go, models.annotations_go.c.go_id == models.GO.id)  # Join GO to annotations_go\n",
    "        .join(models.Annotation, models.Annotation.id == models.annotations_go.c.annotation_id)  # Join annotations_go to annotations\n",
    "        .join(models.Gene, models.Gene.id == models.Annotation.gene_id)  # Join annotations to genes\n",
    "        .join(models.Species, models.Species.id == models.Gene.species_id)  # Join genes to species\n",
    "        .group_by(models.GO.go_id, models.GO.go_name)  # Group by GO ID and name\n",
    "        .having(func.count(models.Gene.id) > threshold)  # Filter by gene count greater than 0\n",
    "        .order_by(func.count(models.Gene.id).asc())  # Order by gene count in ascending order\n",
    "        .filter(models.Species.name == \"X. elegans\") \n",
    "    )\n",
    "\n",
    "    return query.all()\n",
    "\n",
    "print(get_go_terms_with_fewest_genes(13))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 24-01-2025 Add Humilis Fasta data and neaten up interface\n",
    "- add humilis species, data\n",
    "- TODO make adding fasta info more efficient \n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the species and gene sequences to the database\n",
    "\n",
    "import db as db\n",
    "import db_manager as dbm\n",
    "\n",
    "\n",
    "\n",
    "database = db.DB()\n",
    "species_name = \"X. humilis\"\n",
    "fasta_file = \"all_data/Xhumilis_Nov2024/Xhum_CDS_annot150424.fasta\"\n",
    "\n",
    "species = database.add_species(species_name)\n",
    "species_id = species.id \n",
    "# dbm.add_gene_sequence_from_fasta(fasta_file, species_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the gene annotations to the database\n",
    "\n",
    "annotation_file = \"all_data/Xhumilis_Nov2024/20241111_Xhumilis_annotation_76405_export table_Oliver.csv\"\n",
    "dbm.add_gene_annotations(annotation_file, species_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import db as db\n",
    "from sqlalchemy.orm import aliased\n",
    "from sqlalchemy.sql import func\n",
    "import models\n",
    "def get_counts_by_species(species_id):\n",
    "    # Count GO terms\n",
    "    database = db.DB()\n",
    "    session = database.session\n",
    "    go_count = session.query(func.count(models.GO.id)) \\\n",
    "        .join(models.annotations_go) \\\n",
    "        .join(models.Annotation) \\\n",
    "        .join(models.Gene) \\\n",
    "        .filter(models.Gene.species_id == species_id) \\\n",
    "        .scalar()\n",
    "\n",
    "    # Count InterPro terms\n",
    "    interpro_count = session.query(func.count(models.InterPro.id)) \\\n",
    "        .join(models.annotations_interpro) \\\n",
    "        .join(models.Annotation) \\\n",
    "        .join(models.Gene) \\\n",
    "        .filter(models.Gene.species_id == species_id) \\\n",
    "        .scalar()\n",
    "\n",
    "    # Count gene annotations\n",
    "    annotation_count = session.query(func.count(models.Annotation.id)) \\\n",
    "        .join(models.Gene) \\\n",
    "        .filter(models.Gene.species_id == species_id) \\\n",
    "        .scalar()\n",
    "\n",
    "\n",
    "    return {\n",
    "        \"GO_terms_count\": go_count,\n",
    "        \"InterPro_count\": interpro_count,\n",
    "        \"Gene_annotations_count\": annotation_count\n",
    "    }\n",
    "\n",
    "# Example usage\n",
    "for i in range(1, 4):\n",
    "    counts = get_counts_by_species(i)\n",
    "    print(counts)\n",
    "\n",
    "293425 +312558 +1017868 = 1,623,851"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 14-02-2025 Add X elegans arabidopsis genes to database\n",
    "\n",
    "- Below code takes in a CSV file with three columns with names [\"Gene name\",\"At Locus ID\",\"Wiki gene description\"]\n",
    "- It then converts it into a dataframe, passes it to DB.py where the At locus and Wiki gene description (common name) are added to the database\n",
    "- the below function has also been added to the db_manager.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 0 gene mappings.\n",
      "Processed 100 gene mappings.\n",
      "Processed 200 gene mappings.\n",
      "Processed 300 gene mappings.\n",
      "Processed 400 gene mappings.\n",
      "Processed 500 gene mappings.\n",
      "Processed 600 gene mappings.\n",
      "Processed 700 gene mappings.\n",
      "Processed 800 gene mappings.\n",
      "Processed 900 gene mappings.\n",
      "Processed 1000 gene mappings.\n",
      "Processed 1100 gene mappings.\n",
      "Processed 1200 gene mappings.\n",
      "Processed 1300 gene mappings.\n",
      "Processed 1400 gene mappings.\n",
      "Processed 1500 gene mappings.\n",
      "Processed 1600 gene mappings.\n",
      "Processed 1700 gene mappings.\n",
      "Processed 1800 gene mappings.\n",
      "Processed 1900 gene mappings.\n",
      "Processed 2000 gene mappings.\n",
      "Processed 2100 gene mappings.\n",
      "Processed 2200 gene mappings.\n",
      "Processed 2300 gene mappings.\n",
      "Processed 2400 gene mappings.\n",
      "Processed 2500 gene mappings.\n",
      "Processed 2600 gene mappings.\n",
      "Processed 2700 gene mappings.\n",
      "Processed 2800 gene mappings.\n",
      "Processed 2900 gene mappings.\n",
      "Processed 3000 gene mappings.\n",
      "Processed 3100 gene mappings.\n",
      "Processed 3200 gene mappings.\n",
      "Processed 3300 gene mappings.\n",
      "Processed 3400 gene mappings.\n",
      "Processed 3500 gene mappings.\n",
      "Processed 3600 gene mappings.\n",
      "Processed 3700 gene mappings.\n",
      "Processed 3800 gene mappings.\n",
      "Processed 3900 gene mappings.\n",
      "Processed 4000 gene mappings.\n",
      "Processed 4100 gene mappings.\n",
      "Processed 4200 gene mappings.\n",
      "Processed 4300 gene mappings.\n",
      "Processed 4400 gene mappings.\n",
      "Processed 4500 gene mappings.\n",
      "Processed 4600 gene mappings.\n",
      "Processed 4700 gene mappings.\n",
      "Processed 4800 gene mappings.\n",
      "Processed 4900 gene mappings.\n",
      "Processed 5000 gene mappings.\n",
      "Processed 5100 gene mappings.\n",
      "Processed 5200 gene mappings.\n",
      "Processed 5300 gene mappings.\n",
      "Processed 5400 gene mappings.\n",
      "Processed 5500 gene mappings.\n",
      "Processed 5600 gene mappings.\n",
      "Processed 5700 gene mappings.\n",
      "Processed 5800 gene mappings.\n",
      "Processed 5900 gene mappings.\n",
      "Processed 6000 gene mappings.\n",
      "Processed 6100 gene mappings.\n",
      "Processed 6200 gene mappings.\n",
      "Processed 6300 gene mappings.\n",
      "Processed 6400 gene mappings.\n",
      "Processed 6500 gene mappings.\n",
      "Processed 6600 gene mappings.\n",
      "Processed 6700 gene mappings.\n",
      "Processed 6800 gene mappings.\n",
      "Processed 6900 gene mappings.\n",
      "Processed 7000 gene mappings.\n",
      "Processed 7100 gene mappings.\n",
      "Processed 7200 gene mappings.\n",
      "Processed 7300 gene mappings.\n",
      "Processed 7400 gene mappings.\n",
      "Processed 7500 gene mappings.\n",
      "Processed 7600 gene mappings.\n",
      "Processed 7700 gene mappings.\n",
      "Processed 7800 gene mappings.\n",
      "Processed 7900 gene mappings.\n",
      "Processed 8000 gene mappings.\n",
      "Processed 8100 gene mappings.\n",
      "Processed 8200 gene mappings.\n",
      "Processed 8300 gene mappings.\n",
      "Processed 8400 gene mappings.\n",
      "Processed 8500 gene mappings.\n",
      "Processed 8600 gene mappings.\n",
      "Processed 8700 gene mappings.\n",
      "Processed 8800 gene mappings.\n",
      "Processed 8900 gene mappings.\n",
      "Processed 9000 gene mappings.\n",
      "Processed 9100 gene mappings.\n",
      "Processed 9200 gene mappings.\n",
      "Processed 9300 gene mappings.\n",
      "Processed 9400 gene mappings.\n",
      "Processed 9500 gene mappings.\n",
      "Processed 9600 gene mappings.\n",
      "Processed 9700 gene mappings.\n",
      "Processed 9800 gene mappings.\n",
      "Processed 9900 gene mappings.\n",
      "Processed 10000 gene mappings.\n",
      "Processed 10100 gene mappings.\n",
      "Processed 10200 gene mappings.\n",
      "Processed 10300 gene mappings.\n",
      "Processed 10400 gene mappings.\n",
      "Processed 10500 gene mappings.\n",
      "Processed 10600 gene mappings.\n",
      "Processed 10700 gene mappings.\n",
      "Processed 10800 gene mappings.\n",
      "Processed 10900 gene mappings.\n",
      "Processed 11000 gene mappings.\n",
      "Processed 11100 gene mappings.\n",
      "Processed 11200 gene mappings.\n",
      "Processed 11300 gene mappings.\n",
      "Processed 11400 gene mappings.\n",
      "Processed 11500 gene mappings.\n",
      "Processed 11600 gene mappings.\n",
      "Processed 11700 gene mappings.\n",
      "Processed 11800 gene mappings.\n",
      "Processed 11900 gene mappings.\n",
      "Processed 12000 gene mappings.\n",
      "Processed 12100 gene mappings.\n",
      "Processed 12200 gene mappings.\n",
      "Processed 12300 gene mappings.\n",
      "Processed 12400 gene mappings.\n",
      "Processed 12500 gene mappings.\n",
      "Processed 12600 gene mappings.\n",
      "Processed 12700 gene mappings.\n",
      "Processed 12800 gene mappings.\n",
      "Processed 12900 gene mappings.\n",
      "Processed 13000 gene mappings.\n",
      "Processed 13100 gene mappings.\n",
      "Processed 13200 gene mappings.\n",
      "Processed 13300 gene mappings.\n",
      "Processed 13400 gene mappings.\n",
      "Processed 13500 gene mappings.\n",
      "Processed 13600 gene mappings.\n",
      "Processed 13700 gene mappings.\n",
      "Processed 13800 gene mappings.\n",
      "Processed 13900 gene mappings.\n",
      "Processed 14000 gene mappings.\n",
      "Processed 14100 gene mappings.\n",
      "Processed 14200 gene mappings.\n",
      "Processed 14300 gene mappings.\n",
      "Processed 14400 gene mappings.\n",
      "Processed 14500 gene mappings.\n",
      "Processed 14600 gene mappings.\n",
      "Processed 14700 gene mappings.\n",
      "Processed 14800 gene mappings.\n",
      "Processed 14900 gene mappings.\n",
      "Processed 15000 gene mappings.\n",
      "Processed 15100 gene mappings.\n",
      "Processed 15200 gene mappings.\n",
      "Processed 15300 gene mappings.\n",
      "Processed 15400 gene mappings.\n",
      "Processed 15500 gene mappings.\n",
      "Processed 15600 gene mappings.\n",
      "Processed 15700 gene mappings.\n",
      "Processed 15800 gene mappings.\n",
      "Processed 15900 gene mappings.\n",
      "Processed 16000 gene mappings.\n",
      "Processed 16100 gene mappings.\n",
      "Processed 16200 gene mappings.\n",
      "Processed 16300 gene mappings.\n",
      "Processed 16400 gene mappings.\n",
      "Processed 16500 gene mappings.\n",
      "Processed 16600 gene mappings.\n",
      "Processed 16700 gene mappings.\n",
      "Processed 16800 gene mappings.\n",
      "Processed 16900 gene mappings.\n",
      "Processed 17000 gene mappings.\n",
      "Processed 17100 gene mappings.\n",
      "Processed 17200 gene mappings.\n",
      "Processed 17300 gene mappings.\n",
      "Processed 17400 gene mappings.\n",
      "Processed 17500 gene mappings.\n",
      "Processed 17600 gene mappings.\n",
      "Processed 17700 gene mappings.\n",
      "Processed 17800 gene mappings.\n",
      "Processed 17900 gene mappings.\n",
      "Processed 18000 gene mappings.\n",
      "Processed 18100 gene mappings.\n",
      "Processed 18200 gene mappings.\n",
      "Processed 18300 gene mappings.\n",
      "Processed 18400 gene mappings.\n",
      "Processed 18500 gene mappings.\n",
      "Processed 18600 gene mappings.\n",
      "Processed 18700 gene mappings.\n",
      "Processed 18800 gene mappings.\n",
      "Processed 18900 gene mappings.\n",
      "Processed 19000 gene mappings.\n",
      "Processed 19100 gene mappings.\n",
      "Processed 19200 gene mappings.\n",
      "Processed 19300 gene mappings.\n",
      "Processed 19400 gene mappings.\n",
      "Processed 19500 gene mappings.\n",
      "Processed 19600 gene mappings.\n",
      "Processed 19700 gene mappings.\n",
      "Processed 19800 gene mappings.\n",
      "Processed 19900 gene mappings.\n",
      "Processed 20000 gene mappings.\n",
      "Processed 20100 gene mappings.\n",
      "Processed 20200 gene mappings.\n",
      "Processed 20300 gene mappings.\n",
      "Processed 20400 gene mappings.\n",
      "Processed 20500 gene mappings.\n",
      "Processed 20600 gene mappings.\n",
      "Processed 20700 gene mappings.\n",
      "Processed 20800 gene mappings.\n",
      "Processed 20900 gene mappings.\n",
      "Processed 21000 gene mappings.\n",
      "Processed 21100 gene mappings.\n",
      "Processed 21200 gene mappings.\n",
      "Processed 21300 gene mappings.\n",
      "Processed 21400 gene mappings.\n",
      "Processed 21500 gene mappings.\n",
      "Processed 21600 gene mappings.\n",
      "Processed 21700 gene mappings.\n",
      "Processed 21800 gene mappings.\n",
      "Processed 21900 gene mappings.\n",
      "Processed 22000 gene mappings.\n",
      "Processed 22100 gene mappings.\n",
      "Processed 22200 gene mappings.\n",
      "Processed 22300 gene mappings.\n",
      "Processed 22400 gene mappings.\n",
      "Processed 22500 gene mappings.\n",
      "Processed 22600 gene mappings.\n",
      "Processed 22700 gene mappings.\n",
      "Processed 22800 gene mappings.\n",
      "Processed 22900 gene mappings.\n",
      "Processed 23000 gene mappings.\n",
      "Processed 23100 gene mappings.\n",
      "Processed 23200 gene mappings.\n",
      "Processed 23300 gene mappings.\n",
      "Processed 23400 gene mappings.\n",
      "Processed 23500 gene mappings.\n",
      "Processed 23600 gene mappings.\n",
      "Processed 23700 gene mappings.\n",
      "Processed 23800 gene mappings.\n",
      "Processed 23900 gene mappings.\n",
      "Processed 24000 gene mappings.\n",
      "Processed 24100 gene mappings.\n",
      "Processed 24200 gene mappings.\n",
      "Processed 24300 gene mappings.\n",
      "Processed 24400 gene mappings.\n",
      "Processed 24500 gene mappings.\n",
      "Processed 24600 gene mappings.\n",
      "Processed 24700 gene mappings.\n",
      "Processed 24800 gene mappings.\n",
      "Processed 24900 gene mappings.\n",
      "Processed 25000 gene mappings.\n",
      "Processed 25100 gene mappings.\n",
      "Processed 25200 gene mappings.\n",
      "Processed 25300 gene mappings.\n",
      "Processed 25400 gene mappings.\n",
      "Processed 25500 gene mappings.\n",
      "Processed 25600 gene mappings.\n",
      "Processed 25700 gene mappings.\n",
      "Processed 25800 gene mappings.\n",
      "Processed 25900 gene mappings.\n",
      "Processed 26000 gene mappings.\n",
      "Processed 26100 gene mappings.\n",
      "Processed 26200 gene mappings.\n",
      "Processed 26300 gene mappings.\n",
      "Processed 26400 gene mappings.\n",
      "Processed 26500 gene mappings.\n",
      "Processed 26600 gene mappings.\n",
      "Processed 26700 gene mappings.\n",
      "Processed 26800 gene mappings.\n",
      "Processed 26900 gene mappings.\n",
      "Processed 27000 gene mappings.\n",
      "Processed 27100 gene mappings.\n",
      "Processed 27200 gene mappings.\n",
      "Processed 27300 gene mappings.\n",
      "Processed 27400 gene mappings.\n",
      "Processed 27500 gene mappings.\n",
      "Processed 27600 gene mappings.\n",
      "Processed 27700 gene mappings.\n",
      "Processed 27800 gene mappings.\n",
      "Processed 27900 gene mappings.\n",
      "Added 27919 gene mappings to the database.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import database.db as db\n",
    "\n",
    "def add_a_thaliana_gene_mapping(mapping_file):\n",
    "    database = db.DB()\n",
    "    \n",
    "    data = pd.read_csv(mapping_file)\n",
    "    \n",
    "    # Ensure columns exist\n",
    "    required_columns = [\"Gene name\",\"At Locus ID\",\"Wiki gene description\"]\n",
    "    if not all(col in data.columns for col in required_columns):\n",
    "        raise ValueError(f\"CSV is missing one or more required columns: {required_columns}\")\n",
    "\n",
    "    data.sort_values(\"At Locus ID\", inplace=True, ascending=False)\n",
    "\n",
    "    database.add_a_thaliana_gene_mappings(data)\n",
    "    # add at common name\n",
    "    # add locus\n",
    "    # make all lower case\n",
    "    # retreive gene ID, \n",
    "\n",
    "mapping_file = \"all_data/arab_gene_mapping/20250212_Xelegans_Arabidopsis_topblast_TAIRID_Mapping.csv\"\n",
    "add_a_thaliana_gene_mapping(mapping_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 19-02-2025 Retrieve all annotation data for a list of genes (At homologue locus, common name, description)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retrieve annotation data from xerophyta gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "import database.db as db\n",
    "\n",
    "database = db.DB()\n",
    "gene_name = [\"Xele.ptg000049l.138\",\"Xele.ptg000049l.140\"]\n",
    "# gene_name = [\"Xele.ptg000049l.140\"]\n",
    "\n",
    "genes = database.get_gene_annotation_data(gene_name, \"xerophyta_gene_name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = []\n",
    "for gene in genes:\n",
    "    species = gene.species.name \n",
    "    gene_name=gene.gene_name\n",
    "    coding_sequence = gene.coding_sequence\n",
    "\n",
    "    for homologue in gene.arabidopsis_homologues:\n",
    "        a_thaliana_locus = homologue.a_thaliana_locus\n",
    "        a_thaliana_common_name = homologue.a_thaliana_common_name\n",
    "\n",
    "    for annotation in gene.annotations:\n",
    "        description = annotation.description\n",
    "        e_value = annotation.e_value\n",
    "        bit_score = annotation.bit_score\n",
    "        similarity = annotation.similarity\n",
    "        alignment_length = annotation.alignment_length\n",
    "        positives = annotation.positives\n",
    "        \n",
    "        go_ids = \", \".join([go.go_id for go in annotation.go_ids])\n",
    "        go_names = \", \".join([go.go_name for go in annotation.go_ids])\n",
    "\n",
    "        enzyme_codes = \", \".join([enzyme_codes.enzyme_code for enzyme_codes in annotation.enzyme_codes])\n",
    "        enzyme_names = \", \".join([enzyme_codes.enzyme_name for enzyme_codes in annotation.enzyme_codes])\n",
    "       \n",
    "        interpro_ids = \", \".join([interpro.interpro_id for interpro in annotation.interpro_ids])\n",
    "        data.append({\n",
    "            \"species\": species,\n",
    "            \"gene_name\": gene_name,\n",
    "            \"coding_sequence\": coding_sequence,\n",
    "            \"a_thaliana_locus\": a_thaliana_locus,\n",
    "            \"a_thaliana_common_name\": a_thaliana_common_name,\n",
    "            \"description\": description,\n",
    "            \"e_value\": e_value,\n",
    "            \"bit_score\": bit_score,\n",
    "            \"similarity\": similarity,\n",
    "            \"alignment_length\": alignment_length,\n",
    "            \"positives\": positives,\n",
    "            \"go_ids\": go_ids,\n",
    "            \"go_names\": go_names,\n",
    "            \"enzyme_codes\": enzyme_codes,\n",
    "            \"enzyme_names\": enzyme_names,\n",
    "            \"interpro_ids\": interpro_ids\n",
    "        })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'species': 'X. elegans', 'gene_name': 'Xele.ptg000049l.138', 'a_thaliana_locus': 'AT3G13850', 'a_thaliana_common_name': 'LOB domain-containing protein 22', 'description': 'RecName: Full=LOB domain-containing protein 22; AltName: Full=ASYMMETRIC LEAVES 2-like protein 30; Short=AS2-like protein 30', 'e_value': 3.75e-33, 'bit_score': None, 'similarity': None, 'alignment_length': None, 'positives': None, 'go_ids': 'P:GO:0009556, P:GO:0009786, P:GO:0009799, P:GO:0009944, P:GO:0009954, P:GO:0009965, P:GO:0010089, P:GO:0010199, P:GO:0010311, P:GO:0010628, P:GO:0045893, P:GO:0048441, P:GO:1905177, P:GO:1990110, F:GO:0000976, F:GO:0005515, C:GO:0005654', 'go_names': 'P:microsporogenesis, P:regulation of asymmetric cell division, P:specification of symmetry, P:polarity specification of adaxial/abaxial axis, P:proximal/distal pattern formation, P:leaf morphogenesis, P:xylem development, P:organ boundary specification between lateral organs and the meristem, P:lateral root formation, P:positive regulation of gene expression, P:positive regulation of DNA-templated transcription, P:petal development, P:tracheary element differentiation, P:callus formation, F:transcription cis-regulatory region binding, F:protein binding, C:nucleoplasm', 'enzyme_codes': '', 'enzyme_names': '', 'interpro_ids': 'IPR004883 (PFAM), PTHR31301 (PANTHER), IPR004883 (PROSITE_PROFILES)'}\n",
      "{'species': 'X. elegans', 'gene_name': 'Xele.ptg000049l.140', 'a_thaliana_locus': 'AT1G17650', 'a_thaliana_common_name': 'glyoxylate reductase 2', 'description': 'RecName: Full=Glyoxylate/succinic semialdehyde reductase 2, chloroplastic; Short=AtGLYR2; Short=AtGR2; Short=SSA reductase 2; Flags: Precursor', 'e_value': 7.04e-153, 'bit_score': None, 'similarity': None, 'alignment_length': None, 'positives': None, 'go_ids': 'P:GO:0006974, P:GO:0007549, P:GO:0009436, P:GO:0009442, P:GO:0019579, P:GO:0045815, P:GO:0140673, F:GO:0003677, F:GO:0008679, F:GO:0030267, F:GO:0031491, F:GO:0035064, F:GO:0050661, F:GO:0051287, F:GO:0140463, C:GO:0000786, C:GO:0005654, C:GO:0005705, C:GO:0005829, C:GO:0009570', 'go_names': 'P:DNA damage response, P:sex-chromosome dosage compensation, P:glyoxylate catabolic process, P:allantoin assimilation pathway, P:aldaric acid catabolic process, P:transcription initiation-coupled chromatin remodeling, P:transcription elongation-coupled chromatin remodeling, F:DNA binding, F:2-hydroxy-3-oxopropionate reductase activity, F:glyoxylate reductase (NADPH) activity, F:nucleosome binding, F:methylated histone binding, F:NADP binding, F:NAD binding, F:chromatin-protein adaptor activity, C:nucleosome, C:nucleoplasm, C:polytene chromosome interband, C:cytosol, C:chloroplast stroma', 'enzyme_codes': 'EC:1.1.1.79, EC:1.1.1.60', 'enzyme_names': 'glyoxylate reductase (NADP(+)), 2-hydroxy-3-oxopropionate reductase', 'interpro_ids': 'G3DSA:3.40.50.720:FF:000058 (FUNFAM), G3DSA:1.10.1040.10:FF:000016 (FUNFAM), IPR006115 (PFAM), IPR013328 (G3DSA:1.10.1040.GENE3D), IPR015815 (PIRSF), IPR029154 (PFAM), G3DSA:3.40.50.720 (GENE3D), PIRSR000114-3 (PIRSR), PIRSR000103-1 (PIRSR), PIRSR000193-1 (PIRSR), PIRSR000445-3 (PIRSR), IPR051265 (PANTHER), IPR002204 (PROSITE_PATTERNS), IPR036291 (SUPERFAMILY), IPR008927 (SUPERFAMILY)'}\n"
     ]
    }
   ],
   "source": [
    "for d in data:\n",
    "    print(d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Retreive annotation data from Arabidopsis homologue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT3G13850 Xele.ptg000018l.1938 \n",
      "AT3G13850 Xele.ptg000028l.172 \n",
      "AT3G13850 Xele.ptg000049l.138 \n",
      "AT1G17650 Xele.ptg000049l.140 \n",
      "{'at3g13850', 'at1g17650'}\n",
      "{'not in'}\n"
     ]
    }
   ],
   "source": [
    "import database.db as db\n",
    "\n",
    "database = db.DB()\n",
    "\n",
    "gene_list = [\"AT1G17650\", \"AT3G13850\", \"not in\"]\n",
    "genes = database.get_gene_annotation_data_from_a_thaliana_homologue(gene_list)\n",
    "for g in genes:\n",
    "    print(f'{g.arabidopsis_homologues[0].a_thaliana_locus} {g.gene_name} ')\n",
    "\n",
    "matched_loci = {gene.arabidopsis_homologues[0].a_thaliana_locus.lower() for gene in genes}\n",
    "print(matched_loci)\n",
    "missing_loci = {gene for gene in gene_list if gene.lower() not in matched_loci}\n",
    "print(missing_loci)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4159\n"
     ]
    }
   ],
   "source": [
    "print(len(genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 23-02-2025 Missing annotation\n",
    "\n",
    "- genes without blast hits dont have an associated annotation object, causing issues when retreiving annotation data\n",
    "- TODO\n",
    "  - confirm which genes dont have associated annotation object\n",
    "  - check which annotation file contains the gene name or not\n",
    "  - 27920 genes in annotaition file but 28572 genes in FASTA/database\n",
    "  - assume the missing 652 arent annotated? \n",
    "    - they wont have an At match because the AT matches only has the  27920 but some of the 27920 genes (4159 genes) also have no blast hit\n",
    "  - CHECK how many Xele genes have annotation in DB i.e select annotation count where species.id = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import database.db as db \n",
    "\n",
    "database = db.DB()\n",
    "x = \"Xele.ptg000041l.177\" #no annotation\n",
    "y =\"Xele.ptg000049l.139\" # no annotation\n",
    "gene = database.get_gene_by_name(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no blast hit\n",
      "4641\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'Gene' object has no attribute 'go'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28mprint\u001b[39m(gene\u001b[38;5;241m.\u001b[39marabidopsis_homologues[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39ma_thaliana_locus)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(gene\u001b[38;5;241m.\u001b[39mid)\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mgene\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgo\u001b[49m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Gene' object has no attribute 'go'"
     ]
    }
   ],
   "source": [
    "print(gene.arabidopsis_homologues[0].a_thaliana_locus)\n",
    "print(gene.id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import database.models as  models\n",
    "\n",
    "database.session.query(models.Gene).where(species).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['no blast hit', '3 or 4', '1', '2', '5']\n"
     ]
    }
   ],
   "source": [
    "def parse_multi_input(text_input):\n",
    "    \"\"\"\n",
    "    Splits the user's input (comma, space, newline) into a list of unique, non-empty strings.\n",
    "    \"\"\"\n",
    "    if not text_input.strip():\n",
    "        return []\n",
    "    # Replace commas/newlines with spaces\n",
    "    cleaned = text_input.replace(\"\\n\", \",\")\n",
    "    # Split on whitespace\n",
    "    tokens = [t.strip() for t in cleaned.split(\",\") if t.strip()]\n",
    "    # Return unique tokens\n",
    "    return list(set(tokens))\n",
    "\n",
    "input = \"no blast hit, 1  , 2, 3 or 4 \\n 5\"\n",
    "print(parse_multi_input(input))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "xerophyta_data_app",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
