{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "file_path = \"all_data/Xschlechteri_Nov2024/ncbi_dataset/ncbi_dataset/data/data_report.jsonl\"\n",
    "\n",
    "data = []\n",
    "\n",
    "\n",
    "with open(file_path, 'r', encoding='utf-8') as file:\n",
    "    for line_number, line in enumerate(file, start=1):\n",
    "        try:\n",
    "            # Parse the JSON object from each line\n",
    "            json_obj = json.loads(line.strip())\n",
    "            data.append(json_obj)\n",
    "        except json.JSONDecodeError as e:\n",
    "            print(f\"Error decoding JSON on line {line_number}: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"annotations\": [\n",
      "        {\n",
      "            \"annotationName\": \"Annotation submitted by TAIR and Araport\",\n",
      "            \"annotationReleaseDate\": \"2022-10-20\",\n",
      "            \"assemblyAccession\": \"GCF_000001735.4\",\n",
      "            \"assemblyName\": \"TAIR10.1\",\n",
      "            \"genomicLocations\": [\n",
      "                {\n",
      "                    \"genomicAccessionVersion\": \"NC_003071.7\",\n",
      "                    \"genomicRange\": {\n",
      "                        \"begin\": \"11678305\",\n",
      "                        \"end\": \"11679919\",\n",
      "                        \"orientation\": \"minus\"\n",
      "                    },\n",
      "                    \"sequenceName\": \"2\"\n",
      "                }\n",
      "            ]\n",
      "        }\n",
      "    ],\n",
      "    \"chromosomes\": [\n",
      "        \"2\"\n",
      "    ],\n",
      "    \"commonName\": \"thale cress\",\n",
      "    \"description\": \"FAM210B-like protein, putative (DUF1279)\",\n",
      "    \"geneId\": \"817272\",\n",
      "    \"geneOntology\": {\n",
      "        \"cellularComponents\": [\n",
      "            {\n",
      "                \"evidenceCode\": \"HDA\",\n",
      "                \"goId\": \"GO:0009507\",\n",
      "                \"name\": \"chloroplast\",\n",
      "                \"qualifier\": \"located_in\"\n",
      "            },\n",
      "            {\n",
      "                \"evidenceCode\": \"ISM\",\n",
      "                \"goId\": \"GO:0009507\",\n",
      "                \"name\": \"chloroplast\",\n",
      "                \"qualifier\": \"located_in\"\n",
      "            },\n",
      "            {\n",
      "                \"evidenceCode\": \"HDA\",\n",
      "                \"goId\": \"GO:0009535\",\n",
      "                \"name\": \"chloroplast thylakoid membrane\",\n",
      "                \"qualifier\": \"located_in\"\n",
      "            },\n",
      "            {\n",
      "                \"evidenceCode\": \"IDA\",\n",
      "                \"goId\": \"GO:0009579\",\n",
      "                \"name\": \"thylakoid\",\n",
      "                \"qualifier\": \"located_in\"\n",
      "            }\n",
      "        ],\n",
      "        \"molecularFunctions\": [\n",
      "            {\n",
      "                \"evidenceCode\": \"ND\",\n",
      "                \"goId\": \"GO:0003674\",\n",
      "                \"name\": \"molecular_function\",\n",
      "                \"qualifier\": \"enables\"\n",
      "            }\n",
      "        ]\n",
      "    },\n",
      "    \"nomenclatureAuthority\": {\n",
      "        \"authority\": \"Araport\",\n",
      "        \"identifier\": \"AT2G27290\"\n",
      "    },\n",
      "    \"orientation\": \"minus\",\n",
      "    \"proteinCount\": 1,\n",
      "    \"symbol\": \"AT2G27290\",\n",
      "    \"synonyms\": [\n",
      "        \"F12K2.13\",\n",
      "        \"F12K2_13\"\n",
      "    ],\n",
      "    \"taxId\": \"3702\",\n",
      "    \"taxname\": \"Arabidopsis thaliana\",\n",
      "    \"transcriptCount\": 1,\n",
      "    \"transcriptTypeCounts\": [\n",
      "        {\n",
      "            \"count\": 1,\n",
      "            \"type\": \"PROTEIN_CODING\"\n",
      "        }\n",
      "    ],\n",
      "    \"type\": \"PROTEIN_CODING\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(data[2], indent=4))  # Pretty-print the first JSON object\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01-01-2025 Add RNAseq data to DB\n",
    "Need to add the X. elegans RNA-seq data to the db. \n",
    "\n",
    "Steps:\n",
    "1. Add tables to `models.py`  and add info to DB schema with Alembic  \n",
    "2. Format data into long data format  \n",
    "\n",
    "### 1. Add table\n",
    "**Set up alembic** \n",
    "\n",
    "`alembic init alembic`  \n",
    "... chatGPT for other steps  \n",
    "\n",
    "**Create a migration script**   \n",
    "\n",
    "`alembic revision --autogenerate -m \"Your migration description here\"`  \n",
    "fill in script\n",
    "`alembic upgrade head`  \n",
    "\n",
    "\n",
    "Added two tables: `Gene_expressions` and `Experiments`\n",
    "\n",
    "### 2. Format the data\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'data_tidier' has no attribute 'transform_to_long'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m file_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall_data/Michael_RNAseq/Xe_seedlings (updated)/Xe_seedlings DESeq2 normalised counts table.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m  \u001b[38;5;66;03m# Replace with your file path\u001b[39;00m\n\u001b[1;32m      5\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(file_path)\n\u001b[0;32m----> 6\u001b[0m long_df \u001b[38;5;241m=\u001b[39m \u001b[43mdt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtransform_to_long\u001b[49m(df)\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28mprint\u001b[39m(long_df\u001b[38;5;241m.\u001b[39mhead())\n",
      "\u001b[0;31mAttributeError\u001b[0m: module 'data_tidier' has no attribute 'transform_to_long'"
     ]
    }
   ],
   "source": [
    "import data_tidier as dt\n",
    "\n",
    "# Load the RNA expression file\n",
    "file_path = \"all_data/Michael_RNAseq/Xe_seedlings (updated)/Xe_seedlings DESeq2 normalised counts table.csv\"  # Replace with your file path\n",
    "df = pd.read_csv(file_path)\n",
    "long_df = dt.transform_to_long(df)\n",
    "\n",
    "print(long_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            gene_name species treatment replicate time  expression\n",
      "0   Xele.ptg000001l.1      Xe        De        R2  T00   24.091222\n",
      "1   Xele.ptg000001l.1      Xe        De        R2  T03   21.161189\n",
      "2   Xele.ptg000001l.1      Xe        De        R2  T06   10.594839\n",
      "3   Xele.ptg000001l.1      Xe        De        R2  T09   15.733636\n",
      "4   Xele.ptg000001l.1      Xe        De        R2  T12   18.745990\n",
      "5   Xele.ptg000001l.1      Xe        De        R2  T15   24.811838\n",
      "6   Xele.ptg000001l.1      Xe        De        R2  T24    5.936432\n",
      "7   Xele.ptg000001l.1      Xe        De        R3  T00   34.937498\n",
      "8   Xele.ptg000001l.1      Xe        De        R3  T03   11.209777\n",
      "9   Xele.ptg000001l.1      Xe        De        R3  T06   10.344533\n",
      "10  Xele.ptg000001l.1      Xe        De        R3  T09   13.642034\n",
      "11  Xele.ptg000001l.1      Xe        De        R3  T12   20.937272\n",
      "12  Xele.ptg000001l.1      Xe        De        R3  T15   28.694295\n",
      "13  Xele.ptg000001l.1      Xe        De        R3  T24   15.170597\n",
      "14  Xele.ptg000001l.1      Xe        De        R4  T00    9.929903\n",
      "15  Xele.ptg000001l.1      Xe        De        R4  T03    7.377970\n",
      "16  Xele.ptg000001l.1      Xe        De        R4  T06   11.491869\n",
      "17  Xele.ptg000001l.1      Xe        De        R4  T09    8.191377\n",
      "18  Xele.ptg000001l.1      Xe        De        R4  T12   19.234313\n",
      "19  Xele.ptg000001l.1      Xe        De        R4  T15   15.583653\n",
      "20  Xele.ptg000001l.1      Xe        De        R4  T24    3.839474\n",
      "21  Xele.ptg000001l.1      Xe        Re        R2  T00   17.463069\n",
      "22  Xele.ptg000001l.1      Xe        Re        R2  T01    5.612328\n",
      "23  Xele.ptg000001l.1      Xe        Re        R2  T02   34.626361\n",
      "24  Xele.ptg000001l.1      Xe        Re        R2  T04    6.773667\n",
      "25  Xele.ptg000001l.1      Xe        Re        R2  T08   13.927613\n",
      "26  Xele.ptg000001l.1      Xe        Re        R2  T12   28.339235\n",
      "27  Xele.ptg000001l.1      Xe        Re        R2  T48   17.728796\n",
      "28  Xele.ptg000001l.1      Xe        Re        R6  T00    5.991167\n",
      "29  Xele.ptg000001l.1      Xe        Re        R4  T01    1.329064\n",
      "30  Xele.ptg000001l.1      Xe        Re        R3  T02   18.429348\n",
      "31  Xele.ptg000001l.1      Xe        Re        R3  T04   14.752325\n",
      "32  Xele.ptg000001l.1      Xe        Re        R3  T08   11.767674\n",
      "33  Xele.ptg000001l.1      Xe        Re        R3  T12   18.576507\n",
      "34  Xele.ptg000001l.1      Xe        Re        R3  T48    8.274255\n",
      "35  Xele.ptg000001l.1      Xe        Re        R7  T00   27.567755\n",
      "36  Xele.ptg000001l.1      Xe        Re        R5  T01   14.580350\n",
      "37  Xele.ptg000001l.1      Xe        Re        R5  T02   11.914933\n",
      "38  Xele.ptg000001l.1      Xe        Re        R5  T04   14.494214\n",
      "39  Xele.ptg000001l.1      Xe        Re        R5  T08   13.722325\n",
      "40  Xele.ptg000001l.1      Xe        Re        R5  T12   20.439839\n",
      "41  Xele.ptg000001l.1      Xe        Re        R5  T48   17.254742\n",
      "42  Xele.ptg000001l.2      Xe        De        R2  T00  346.704101\n",
      "43  Xele.ptg000001l.2      Xe        De        R2  T03  369.262754\n",
      "44  Xele.ptg000001l.2      Xe        De        R2  T06  346.980973\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mybase",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
